### ISSUE TITLE
Integrating EigenPro3 Model with PyTorch Training Workflow Using Adam Optimizer

### ISSUE THREAD

**Issue by caliber1313** [EigenPro/EigenPro3#5]
Hello, I’m an undergraduate student who has discovered this repository to be a source of inspiration. I hope you won’t mind addressing my question, which may seem a bit silly:  

My question is as follows:
1. Is it possible to train a kernel regression model, similar to EigenPro3, using the Adam optimizer or SGD under autograd of pytorch? 

2. Specifically, can we follow the model formula `prediction = kernel_fn(features, center) @ weights`, where we use `kernel_fn(features, center)` as input (pre-transformed) and focus solely on just training the weights parameters with the Adam optimizer?

3. Additionally, can we transfer the weights trained on EigenPro3 as the initial weights for this weight layer and continue training with Adam and others loss function utilizing the gradient accumulation of pytorch

I really like the concept and want to customize this further within pytorch training workflow (adam, autograd, grad accumulation) environment that I familiar with 

Thank you

**Comment by parthe** [comment]
Hi,

Thanks for your interest.

While something like this might be possible, there are several issues that need to be kept in mind.

1. we cannot store `kernel_fn(features, centers)` in memory.
2. the problem is ill-conditioned so the convergence of Adam may be slow (but this is worth playing around with at - least for small problem sizes).

The transfer thing is certainly possible.

We recommend you go through our paper for more insight into the construction of our algorithm: https://arxiv.org/abs/2302.02605

Good luck with your explorations. Hope you enjoy reading our paper.

Regards,
Parthe
